src/env.rs

```diff
@@ -0,0 +1,296 @@
+    /// of a different type.
+    pub fn set<'a, V: ValueType<'a>>(&mut self, key: Key<V>, value: impl Into<V::Owned>)
+    where
+        V::Owned: Into<Value>,
+    {
+        let env = Arc::make_mut(&mut self.0);
+        let value = value.into().into();
+        let key = key.into();
+        // TODO: use of Entry might be more efficient
+        if let Some(existing) = env.map.get(&key) {
+            if !existing.is_same_type(&value) {
+                panic!(
+                    "Invalid type for key '{}': {:?} differs in kind from {:?}",
+                    key, existing, value
+                );
+            }
+        }
+        env.map.insert(key, value);
+    }
+}
+
+impl Debug for Value {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        match self {
+            Value::Point(p) => write!(f, "Point {:?}", p),
+            Value::Size(s) => write!(f, "Size {:?}", s),
+            Value::Rect(r) => write!(f, "Rect {:?}", r),
+            Value::Color(c) => write!(f, "Color {:?}", c),
+            Value::Float(x) => write!(f, "Float {}", x),
+            Value::String(s) => write!(f, "String {:?}", s),
+        }
+    }
+}
+
+impl<T> Key<T> {
+    pub const fn new(key: &'static str) -> Self {
+        Key {
+            key,
+            value_type: PhantomData,
+        }
+    }
+}
+
+impl Value {
+    pub fn to_inner_unchecked<'a, V: ValueType<'a>>(&'a self) -> V {
+        match ValueType::try_from_value(self) {
+            Ok(v) => v,
+            Err(s) => panic!("{}", s),
+        }
+    }
+
+    fn is_same_type(&self, other: &Value) -> bool {
+        use Value::*;
+        match (self, other) {
+            (Point(_), Point(_)) => true,
+            (Size(_), Size(_)) => true,
+            (Rect(_), Rect(_)) => true,
+            (Color(_), Color(_)) => true,
+            (Float(_), Float(_)) => true,
+            (String(_), String(_)) => true,
+            _ => false,
+        }
+    }
+}
+
+impl Data for Value {
+    fn same(&self, other: &Value) -> bool {
+        use Value::*;
+        match (self, other) {
+            (Point(p1), Point(p2)) => p1.x.same(&p2.x) && p1.y.same(&p2.y),
+            (Rect(r1), Rect(r2)) => {
+                r1.x0.same(&r2.x0) && r1.y0.same(&r2.y0) && r1.x1.same(&r2.x1) && r1.y1.same(&r2.y1)
+            }
+            (Size(s1), Size(s2)) => s1.width.same(&s2.width) && s1.height.same(&s2.height),
+            (Color(c1), Color(c2)) => c1.as_rgba_u32() == c2.as_rgba_u32(),
+            (Float(f1), Float(f2)) => f1.same(&f2),
+            (String(s1), String(s2)) => s1 == s2,
+            _ => false,
+        }
+    }
+}
+
+impl Data for Env {
+    fn same(&self, other: &Env) -> bool {
+        Arc::ptr_eq(&self.0, &other.0) || self.0.deref().same(other.0.deref())
+    }
+}
+
+impl Data for EnvImpl {
+    fn same(&self, other: &EnvImpl) -> bool {
+        if self.map.len() != other.map.len() {
+            return false;
+        }
+        for (k, v1) in self.map.iter() {
```

Comment:
🏌 
```rust
        self.map
            .iter()
            .all(|(k, v1)| other.map.get(k).map(|v2| v1.same(v2)).unwrap_or(false))
```

Reply:
Hmm, I *personally* find the `for` version easier to read, but suspect people will have different feelings about this.

Reply:
either works for me.

Reply:
Heh, I can get rid of all explicit control flow:

```rust
    fn same(&self, other: &EnvImpl) -> bool {
        self.map.len() == other.map.len()
            && self
                .map
                .iter()
                .all(|(k, v1)| other.map.get(k).map(|v2| v1.same(v2)).unwrap_or(false))
    }
```

This isn't bad, actually, I think I'll go with it.